packages/core-data/src/resolvers.js

```diff
@@ -103,7 +103,46 @@ export function* getEmbedPreview( url ) {
  * Requests Upload Permissions from the REST API.
  */
 export function* hasUploadPermissions() {
-	const response = yield apiFetch( { path: '/wp/v2/media', method: 'OPTIONS', parse: false } );
+	yield* canUser( 'create', 'media' );
+}
+
+/**
+ * Checks whether the current user can perform the given action on the given
+ * REST resource.
+ *
+ * @param {string}  action   Action to check. One of: 'create', 'read', 'update',
+ *                           'delete'.
+ * @param {string}  resource REST resource to check, e.g. 'media' or 'posts'.
+ * @param {?string} id       ID of the rest resource to check.
+ */
+export function* canUser( action, resource, id ) {
+	const methods = {
+		create: 'POST',
+		read: 'GET',
+		update: 'PUT',
+		delete: 'DELETE',
+	};
+
+	const method = methods[ action ];
+	if ( ! method ) {
+		throw new Error( `'${ action }' is not a valid action` );
+	}
+
+	const path = id ? `/wp/v2/${ resource }/${ id }` : `/wp/v2/${ resource }`;
+
+	let response;
+	try {
+		response = yield apiFetch( {
+			path,
+			// Ideally this would always be an OPTIONS request, but unfortunately there's
+			// a bug in the REST API which causes the Allow header to not be sent on
+			// OPTIONS requests to /posts/:id routes.
```

Comment:
Is that just for /post/:id of for the all resources? I mentioned that because the logic below happens for all resources.

Is there a ticket/issue covering the bug?

Reply:
I've only checked posts but I suspect the bug affects all REST endpoints. Will do some investigating and open up a Trac ticket 👍

Reply:
If this is the case, it should be fixed /cc @danielbachhuber @kadamwhite

Reply:
If you just want the `Allow` header, what about doing a `HEAD` request instead of `GET`.

Reply:
Looking into this further:

1. The `Allow` header is generated by `rest_send_allow_header()`, which is called on `rest_post_dispatch`.
    * https://github.com/WordPress/WordPress/blob/d2782aabc51f387c405eacc2ddbd38dfdacd4405/wp-includes/rest-api.php#L628
    * https://github.com/WordPress/WordPress/blob/d2782aabc51f387c405eacc2ddbd38dfdacd4405/wp-includes/rest-api.php#L167
2. `rest_post_dispatch` is called for every response (I think):
    * https://github.com/WordPress/WordPress/blob/47d32decd6668b3ccf100acc8a40aa0a9fc35003/wp-includes/rest-api/class-wp-rest-server.php#L352

It seems like `Allow` should be present for `OPTIONS`. It's an officially supported header: https://developer.mozilla.org/en-US/docs/Web/HTTP/Methods/OPTIONS

Can you provide steps to reproduce?

Reply:
I briefly debugged this when I noticed the issue and from memory it was something to do with `rest_handle_options_request` not calling `$request->set_url_params()` unlike `dispatch` which correctly parses and sets URL parameters.

> Can you provide steps to reproduce?

1. Create a reusable block
1. Perform a `GET http://local.wordpress.test/wp-json/wp/v2/blocks` request and note the ID of the reusable block
1. Perform a `OPTIONS http://local.wordpress.test/wp-json/wp/v2/blocks/27` request, replacing `27` with the ID from step 2. Note that an `Allow` header is not sent back
1. Perform a `GET http://local.wordpress.test/wp-json/wp/v2/blocks/27` request, replacing `27` with the ID from step 2. Note that an `Allow` header is sent back

Reply:
I created a ticket here: https://core.trac.wordpress.org/ticket/45753

Reply:
Added a link to the Trac ticket in cf2d7012f.